can/interfaces/socketcan/socketcan.py

```diff
@@ -619,6 +634,7 @@ def __init__(
         self,
         channel: str = "",
         receive_own_messages: bool = False,
+        use_system_timestamp: bool = True,
```

Comment:
The new option should tell about the hardware timestamping and not about the system timestamps which is the default for a good reason: It is always available.
The hardware timestamps are read from some CAN controller chips and its support is not very common.

Instead of `use_system_timestamp` I would suggest to provide a new option:
`can_hardware_timestamps: bool = False,`

So when people really know what they are doing, enabling this feature should be possible.

Reply:
Ill modify to adopt option naming as suggested - makes sense to me.

WRT defaults etc, my initial suggestion of defaulting to SO_TIMESTAMPNS makes sense as I'm a cautious kinda guy. However, reading the kernel docs :
[](https://www.kernel.org/doc/Documentation/networking/timestamping.txt)

I think one should be able to request SOF_TIMESTAMPING_RAW_HARDWARE and either setsockopt() will fail allowing one to fallback to system generated timestamps, or one can use the system timestamp in the cmsg array that's returned.   Annoyingly I don't have enough hardware to test this otherwise Id have had a crack at the same time! Worth thinking about for the future however...

Reply:
To provide the best user experience, the standard (sw) timestamp should be provided in the best resolution available. In whatever way this can be retrieved by the system.

But the hardware timestamping provides values that might not fit to the common system time. And therefore this should be an explicit setting for people that are aware of such details.

Reply:
> But the hardware timestamping provides values that might not fit to the common system time. And therefore this should be an explicit setting for people that are aware of such details.

If referencing system time is helpful, given that the system timestamp and hardware timestamp are both returned in a cmsg array, one could potentially read the system timestamp as well for the first received timestamp and calculate the offset which you apply equally to all subsequent frame timestamps. This would add little overhead but would give you equivalent clock time but with the precision of the hardware timestamp?